modules/framework.lua

```diff
@@ -0,0 +1,1321 @@
+end
+
+function Plugin:report(metrics)
+  self:emit('report')
+  self:onReport(metrics)
+end
+
+function Plugin:onReport(metrics)
+  -- metrics can be { metric = value }
+  -- or {metric = {value, source}}
+  -- or {metric = {{value, source}, {value, source}, {value, source}}
+  -- or {metric, value, source}
+  -- or {{metric, value, source, timestamp}}
+  for metric, v in pairs(metrics) do
+    -- { { metric, value .. }, { metric, value .. } }
+    if type(metric) == 'number' then
+      print(self:format(v.metric, v.value, v.source, v.timestamp or currentTimestamp()))
+    elseif type(v) ~= 'table' then
+      print(self:format(metric, v, self.source, currentTimestamp()))
+    elseif type(v[1]) ~= 'table' and v.value then
+      -- looking for { metric = { value, source, timestamp }}
+      local source = v.source or self.source
+      local value = v.value
+      local timestamp = v.timestamp or currentTimestamp()
+      print(self:format(metric, value, source, timestamp))
+    else
+      -- looking for { metric = {{ value, source, timestamp }}}
+      for _, j in pairs(v) do
+        local source = j.source or self.source
+        local value = j.value
+        local timestamp = j.timestamp or currentTimestamp()
+        print(self:format(metric, value, source, timestamp))
+      end
+    end
+  end
+end
+
+function Plugin:format(metric, value, source, timestamp)
+  self:emit('format')
+  return self:onFormat(metric, value, source, timestamp)
+end
+
+--- Called by the framework before formating the metric output.
+-- @string metric the metric name
+-- @param value the value to format
+-- @string source the source to report for the metric
+-- @param timestamp the time the metric was retrieved
+-- You can override this on your plugin instance.
+function Plugin:onFormat(metric, value, source, timestamp)
+  return string.format('%s %f %s %s', metric, value, source, timestamp)
+end
+
+local CommandPlugin = Plugin:extend()
+framework.CommandPlugin = CommandPlugin
+
+function CommandPlugin:initialize(params)
+  Plugin.initialize(self, params)
+
+  if not params.command then
+    error('params.command undefined. You need to define the command to excetue.')
+  end
+  self.command = params.command
+end
+
+function CommandPlugin:execCommand(callback)
+  local proc = io.popen(self.command, 'r')
+  local output = proc:read("*a")
+  proc:close()
+  if callback then
+    callback(output)
+  end
+end
+
+function CommandPlugin:onPoll()
+  self:execCommand(function (output) self.parseCommandOutput(self, output) end)
+end
+
+function CommandPlugin:parseCommandOutput(output)
+  local metrics = self:onParseCommandOutput(output)
+  self:report(metrics)
+end
+
+function CommandPlugin:onParseCommandOutput(output)
+  print(output)
+  return {}
+end
+
+local NetPlugin = Plugin:extend()
+framework.NetPlugin = NetPlugin
+
+local HttpPlugin = Plugin:extend()
+framework.HttpPlugin = HttpPlugin
+
+function HttpPlugin:initialize(params)
+  Plugin.initialize(self, params)
```

Comment:
@codemoran I've confirmed that Luvit does not check for self-signed certificates. I've configured an nginx instance locally with a self signed certificate and the plugin get the metrics correctly.
```
vagrant@vagrant-ubuntu-trusty-32:/vagrant/boundary/boundary-plugin-nginx$ cat param.json
{
        "pollInterval" : 1000,
        "url" : "https://localhost/nginx_status",
        "source": "nginx_enterprise"
}
```

```
vagrant@vagrant-ubuntu-trusty-32:/vagrant/boundary/boundary-plugin-nginx$ /usr/bin/boundary-meter --lua init.lua
_bevent: Boundary NGINX Plugin version 2.0: Up |t:info|tags:lua,plugin,
NGINX_WAITING 0.000000 nginx_enterprise 1431576755
NGINX_HANDLED 0.000000 nginx_enterprise 1431576755
NGINX_WRITING 1.000000 nginx_enterprise 1431576755
NGINX_REQUESTS_PER_CONNECTION 0.000000 nginx_enterprise 1431576755
NGINX_ACTIVE_CONNECTIONS 1.000000 nginx_enterprise 1431576755
NGINX_REQUESTS 0.000000 nginx_enterprise 1431576755
NGINX_READING 0.000000 nginx_enterprise 1431576755
NGINX_NOT_HANDLED 0.000000 nginx_enterprise 1431576755
```

To also ensure that I've configured it correctly, I did a wget (that by default checks for self-signed certs)

```
vagrant@vagrant-ubuntu-trusty-32:/vagrant/boundary/boundary-plugin-nginx$ wget https://localhost/nginx_stats
--2015-05-14 01:12:53--  https://localhost/nginx_stats
Resolving localhost (localhost)... 127.0.0.1
Connecting to localhost (localhost)|127.0.0.1|:443... connected.
ERROR: cannot verify localhost's certificate, issued by ‘/C=BR/ST=Sao Paulo/O=Internet Widgits Pty Ltd’:
  Self-signed certificate encountered.
```